src/modules/hyprland/workspaces.cpp

```diff
@@ -139,23 +172,95 @@ void Workspaces::remove_workspace(std::string name) {
                                 [&](std::unique_ptr<Workspace> &x) { return x->name() == name; });
 
   if (workspace == workspaces_.end()) {
+    // happens when a workspace on another monitor is destroyed
+    return;
+  }
+
+  if ((*workspace)->is_persistent()) {
     return;
   }
 
   box_.remove(workspace->get()->button());
   workspaces_.erase(workspace);
 }
 
+void Workspaces::fill_persistent_workspaces() {
+  if (config_["persistent_workspaces"].isObject() && !all_outputs()) {
+    const Json::Value persistent_workspaces = config_["persistent_workspaces"];
+    const std::vector<std::string> keys = persistent_workspaces.getMemberNames();
+
+    for (const std::string &key : keys) {
+      const Json::Value &value = persistent_workspaces[key];
+      if (value.isInt()) {
+        // value is a number => create that many workspaces for this monitor
+        // only add if either:
+        // 1. key is "*" and this monitor is not already defined in the config
+        // 2. key is the current monitor name
+        if ((key == "*" && std::find(keys.begin(), keys.end(), bar_.output->name) == keys.end()) ||
+            key == bar_.output->name) {
+          int amount = value.asInt();
+          spdlog::debug("Creating {} persistent workspaces for monitor {}", amount,
+                        bar_.output->name);
+          for (int i = 0; i < amount; i++) {
+            persistent_workspaces_to_create_.emplace_back(
+                std::to_string(monitor_id_ * amount + i + 1));
+          }
+        }
+
+      } else if (value.isArray() && !value.empty()) {
+        // value is an array => key is a workspace name
+        // values are monitor names this workspace should be shown on
+        for (const Json::Value &monitor : value) {
+          if (monitor.isString() && monitor.asString() == bar_.output->name) {
+            persistent_workspaces_to_create_.emplace_back(key);
+            break;
+          }
+        }
+      }
+    }
+  }
+}
+
+void Workspaces::create_persistent_workspaces() {
+  for (const std::string &workspace_name : persistent_workspaces_to_create_) {
+    Json::Value new_workspace;
+    new_workspace["id"] = workspace_name.starts_with("special") ? -99 : 0;
+    new_workspace["name"] = workspace_name;
+    new_workspace["monitor"] = bar_.output->name;
+    new_workspace["windows"] = 0;
+    new_workspace["persistent"] = true;
+
+    create_workspace(new_workspace);
+  }
+}
+
 void Workspaces::init() {
-  active_workspace_name = (gIPC->getSocket1JsonReply("activeworkspace"))["name"].asString();
+  active_workspace_name_ = (gIPC->getSocket1JsonReply("activeworkspace"))["name"].asString();
+
+  // get monitor ID from name (used by persistent workspaces)
+  monitor_id_ = 0;
```

Comment:
why use id here instead of desc?

Reply:
Because I use this ID to calculate the IDs the persistent workspaces should get.

I [forked](https://github.com/zjeffer/split-monitor-workspaces) Duckonaut's `split-monitor-workspaces` Hyprland plugin to make it possible to automatically set up keyboard shortcuts to switch to the xth workspace on the current focused monitor. The plugin calculates the IDs the workspaces should get based on the monitor ID, and here I used the same formula (because the IDs have to match).

Reply:
would make more sense to use ```workspace name:<monitor-desc-here>_<workspace-number-here>```, monitor id can change on restart but monitor desc is persistent.
Also it's more consistent with the rest of hyprland

Reply:
But how can I then ensure both waybar and the plugin give the same name to the workspaces?

Reply:
waybar should only use the name you set in the config file

Reply:
What if you want waybar to work with any monitor you connect, without having to specify the name in the config file?

For example, if I connect my laptop to a monitor it has never seen before, waybar and my fork of the plugin will work together to automatically show persistent workspaces and set up the correct keybinds.

If we did it with monitor descriptions, you would have to do all that stuff manually.

Reply:
use the port name then

Reply:
I don't understand why it would be any different with the port name instead of the monitor description. I would still have to manually specify the port name in the configs.

In [the plugin](https://github.com/zjeffer/split-monitor-workspaces/blob/main/src/main.cpp#L63), I use the formula `workspace_name = (monitor_ID * amount_of_workspaces) + index + 1`. Here, I use the same formula so that any monitor I connect, both the plugin and the monitor will assign the same index to the same workspace ID (with 5 workspaces per monitor, the 4th workspace on the monitor with ID 2 will have workspace ID = 2*5 + 3 + 1 = 15).

I don't see how I can implement the same functionality with port names or monitor desc. I'm open to suggestions.

Reply:
simply use named workspaces, containing the monitor name and a number (```name:<monitor-desc-here>_<workspace-number-here>```)
this way you wouldn't rely on using initial id, which depends on the order you connect the monitors

if you want every monitor you connect on DP-1 (for example) to have the same configuration you would use DP-1 in the config
if you want each monitor you connect on DP-1 (for example) to have a different config you could use the monitor description

using id gives you much less control than either of these options

Reply:
I'm starting to understand. But I think changing it to use monitor descriptions might be more appropriate for another PR, as a lot of things would need to be refactored for it to work. Not to mention the changes I would have to make to the plugin.

I think we can merge this if you agree? CC @Alexays